app/src/org/commcare/activities/LoginActivity.java

```diff
@@ -268,6 +273,17 @@ private void doLogin(LoginMode loginMode, boolean restoreSession, String passwor
         }
     }
 
+    private boolean isUsernameValid(String username) {
+        if (username !=null && !username.isEmpty()) {
+            if (!username.contains("@") || (username.endsWith("@" + HiddenPreferences.getUserDomain()))) {
+                return true;
+            } else {
+                return false;
+            }
+        }
+        return true;

```

Comment:
this should return `false` given username can't be empty.

Reply:
good point. My initial approach was around lack of permission, but then changed to `username` validity. Also changed the order of the checks, the `username` should be first